Ballerina-Training/Ballerina-Forum-Application/back-end/bal-forum/main.bal

```diff
@@ -0,0 +1,211 @@
+        `);
+
+            user.subscribtions.push(post.id);
+            _ = check forumDbClient->execute(`
+                UPDATE users SET subscribtions = ${user.subscribtions.toJsonString()} WHERE id = ${id}
+            `);
+
+            check commit;
+        }
+
+        return {
+            body: {
+                message: "Post created successfully!"
+            }
+        };
+    }
+
+    resource function get posts() returns ForumPost[]|error {
+        stream<ForumPostInDatabase, sql:Error?> postStream = forumDbClient->query(`SELECT * FROM posts`);
+        ForumPost[] posts = check from var post in postStream
+            select check transformPostFromDatabase(post);
+        return posts;
+    }
+
+    resource function post posts/[string id]/like(Like like) returns Ok|NotFound|Conflict|error {
+        ForumPostInDatabase|error postDb = forumDbClient->queryRow(`SELECT * FROM posts WHERE id = ${id}`);
+        if postDb is error {
+            NotFound postNotFound = {
+                body: {
+                    error_message: "Post not found"
+                }
+            };
+            return postNotFound;
+        }
+
+        ForumPost post = check transformPostFromDatabase(postDb);
+        if post.likes.indexOf(like.userId) is int {
+            Conflict alreadyLiked = {
+                body: {
+                    error_message: "You have already liked this post"
+                }
+            };
+            return alreadyLiked;
+        }
+
+        post.likes.push(like.userId);
+        _ = check forumDbClient->execute(`
+                UPDATE posts SET likes = ${post.likes.toJsonString()} WHERE id = ${id}
+            `);
+
+        return {
+            body: {
+                message: "Post liked successfully!"
+            }
+        };
+    }
+
+    resource function get posts/[string id]() returns ForumPost|NotFound|error {
+        ForumPostInDatabase|error postDb = forumDbClient->queryRow(`SELECT * FROM posts WHERE id = ${id}`);
+        if postDb is error {
+            return {
+                body: {
+                    error_message: "Post not found"
+                }
+            };
+        }
+        return transformPostFromDatabase(postDb);
+    }
+
+    resource function post posts/[string id]/comments(NewPostComment newComment) returns Created|NotFound|error {
+        ForumPostInDatabase|error postDb = forumDbClient->queryRow(`SELECT * FROM posts WHERE id = ${id}`);
+        if postDb is error {
+            return {
+                body: {
+                    error_message: "Post not found"
+                }
+            };
+        }
+
+        ForumPost post = check transformPostFromDatabase(postDb);
+        PostComment comment = check createCommentFromNewComment(newComment);
+
+        post.comments.push(comment);
+        _ = check forumDbClient->execute(`
+                UPDATE posts SET comments = ${post.comments.toJsonString()} WHERE id = ${id}
```

Comment:
Wondering if this is the best practice to store multiple items in one cell in a DB. Shouldn't we use a separate table to store the comments along with the post id?

Reply:
Agree, will change accordingly

Reply:
Fixed via [6ebe426](https://github.com/ballerina-guides/training-content/pull/12/commits/6ebe4267085dce1756b9133d627a3d39af4a5bc6)